oracle/oracle.go

```diff
@@ -786,3 +786,92 @@ func nodePoolContainsNode(s []containerengine.Node, e string) bool {
 	}
 	return false
 }
+
+func (o *oracleOps) SetClusterVersion(version string, timeout time.Duration) error {
+	logrus.Println("Setting Cluster version to", version)
+	req := containerengine.UpdateClusterRequest{
+		ClusterId: &o.clusterID,
+		UpdateClusterDetails: containerengine.UpdateClusterDetails{
+			KubernetesVersion: &version,
+		},
+	}
+
+	resp, err := o.containerEngine.UpdateCluster(context.Background(), req)
+	if err != nil {
+		return err
+	}
+
+	return o.waitTillWorkStatusIsSucceeded(resp.OpcRequestId, resp.OpcWorkRequestId, timeout)
+}
+
+func (o *oracleOps) SetInstanceGroupVersion(instanceGroupName string, version string, timeout time.Duration) error {
+	logrus.Println("Setting Instance group version to", version)
+	//get nodepool ID from name
+	var instanceGroupID *string
+	nodePoolsReq := containerengine.ListNodePoolsRequest{CompartmentId: &o.compartmentID, Name: &instanceGroupName, ClusterId: &o.clusterID}
+	nodePools, err := o.containerEngine.ListNodePools(context.Background(), nodePoolsReq)
+	if err != nil {
+		return err
+	}
+
+	if len(nodePools.Items) == 0 {
+		return errors.New("No node pool found with name" + instanceGroupName)
+	}
+	instanceGroupID = nodePools.Items[0].Id
+
+	//update kubernetes version of nodepool
+	resp, err := o.containerEngine.UpdateNodePool(context.Background(), containerengine.UpdateNodePoolRequest{
+		NodePoolId: instanceGroupID,
+		UpdateNodePoolDetails: containerengine.UpdateNodePoolDetails{
+			KubernetesVersion: &version,
+		},
+	})
+
+	if err != nil {
+		return err
+	}
+
+	err = o.waitTillWorkStatusIsSucceeded(resp.OpcRequestId, resp.OpcWorkRequestId, timeout)
+	if err != nil {
+		return err
+	}
+
+	//get existing total node count of instanceGroup
+	existingClusterSize, err := o.GetInstanceGroupSize(instanceGroupName)
+	if err != nil {
+		return err
+	}
+
+	numberOfZones := len(nodePools.Items[0].NodeConfigDetails.PlacementConfigs)
+	totalClusterSize := int(existingClusterSize)
+
+	//get all zones to be updated
+	nodePoolPlacementConfigDetails := make([]containerengine.NodePoolPlacementConfigDetails, numberOfZones)
+
+	for i, placementConfigs := range nodePools.Items[0].NodeConfigDetails.PlacementConfigs {
+		nodePoolPlacementConfigDetails[i].AvailabilityDomain = placementConfigs.AvailabilityDomain
+		nodePoolPlacementConfigDetails[i].SubnetId = placementConfigs.SubnetId
+	}
+	//delete all nodes from existing node pool
+	if err := o.SetInstanceGroupSize(instanceGroupName, 0, timeout); err != nil {
+		return err
+	}
+
+	//create same number of nodes again with updated version
+	req := containerengine.UpdateNodePoolRequest{
```

Comment:
Can we use existing [SetInstanceGroupSize](https://github.com/libopenstorage/cloudops/blob/master/oracle/oracle.go#L476) function to set the size to original size?

Reply:
The SetInstanceGroupSize functions sets the cluster size per domain, we need functionality to create same number of instances as original. If number of instances are not in multiple of number of domains (edge case), the functionality might break. Hence not calling SetInstanceGroupSize here.